src/inmanta/logging.py

```diff
@@ -0,0 +1,300 @@
+    - `log_file_level`: the logging level for the file handler (if `log_file` is set).
+    - `verbose`: the verbosity level of the log messages.
+    - `timed`: if true,  adds the time to the formatter in the log lines.
+
+    The setup is not done in one step as we want logs for the cmd_parser, which will provide the options needed to configure
+    the 'final' logger with `apply_options`.
+
+    for more fine-grained configuration the following functions can be used as well:
+        - `set_log_level`
+        - `set_log_formatter`
+        - `set_logfile_location`
+    """
+
+    _instance: Optional["InmantaLoggerConfig"] = None
+
+    def __init__(self, stream: TextIO = sys.stdout) -> None:
+        """
+        Set up the logging handler for Inmanta
+
+        :param stream: The stream to send log messages to. Default is standard output (sys.stdout).
+        """
+        self._handler: logging.Handler = logging.StreamHandler(stream=stream)
+        self.set_log_level("INFO")
+        formatter = self._get_log_formatter_for_stream_handler(timed=False)
+        self.set_log_formatter(formatter)
+
+        logging.root.handlers = []
+        logging.root.addHandler(self._handler)
+        logging.root.setLevel(0)
+
+    @classmethod
+    @stable_api
+    def get_instance(cls, stream: TextIO = sys.stdout) -> "InmantaLoggerConfig":
+        """
+        This method should be used to obtain an instance of this class, because this class is a singleton.
+
+        :param stream: The stream to send log messages to. Default is standard output (sys.stdout)
+        """
+        if not cls._instance:
+            cls._instance = cls(stream)
+        return cls._instance
+
+    @classmethod
+    @stable_api
+    def clean_instance(cls) -> "InmantaLoggerConfig":
+        """
+        This method should be used to obtain an instance of this class, because this class is a singleton.
+
+        :param stream: The stream to send log messages to. Default is standard output (sys.stdout)
+        """
+        if cls._instance and cls._instance._handler:
+            cls._instance._handler.close()
+        cls._instance = None
+
+    @stable_api
+    def apply_options(self, options: Options) -> None:
+        """
+        Apply the logging options to the current handler. A handler should have been created before
+
+        :param options: the option object coming from the command line. This function use the following
+            attributes: log_file, log_file_level, verbose, timed
+        """
+        if not self._handler:
+            raise Exception("No handler to apply options to. Please use the get_instance method before calling this one")
+        if options.log_file:
+            self.set_logfile_location(options.log_file)
+            formatter = logging.Formatter(fmt="%(asctime)s %(levelname)-8s %(name)-10s %(message)s")
+            self.set_log_formatter(formatter)
+            self.set_log_level(options.log_file_level, cli=False)
```

Comment:
`log_file_level` can be None, but `set_log_level()` accepts `str`?

Reply:
I fixed it by changing
`log_file_level: str = "INFO"
`
so it can't be None anymore. This is also the default value in the CLI